### DISCUSSION TITLE
How to only compute diagonal of exact posterior distribution

### DISCUSSION THREAD

**Discussion by jccalvojackson** [cornellius-gp/gpytorch#2312] [Q&A] [answered]
After doing exact regression I would like to get the predicted variance at some test points-- only the diagonal of the covariance matrix of the posterior at these points.

For this I would do something like this:

```
    def predict(
        self, X: torch.Tensor
    ) -> Tuple[torch.Tensor, torch.Tensor]:
        self.eval()
        self.likelihood.eval()
        with torch.no_grad(), gpytorch.settings.fast_pred_var():
            pred = self.likelihood(self(X))
        return pred.mean, pred.variance
```

But I suspect that `pred` computes the full posterior cov matrix when I only need the variance. Is this True? how could I make this faster?

**Comment by jacobrgardner** [comment, marked as answer]
Nope, you're already doing the right thing! 

By default in GPyTorch, covariance matrices are evaluated and passed around lazily as `LinearOperators`. Thus, `pred.lazy_covariance_matrix` is an object that represents the full covariance matrix, but doesn't consume the memory or compute time required to build the whole thing. Calling `pred.variance` will only get you the diagonal. It's only if you start asking for e.g. off diagonal elements or functions of the covariance matrix that require the whole thing that we start building covariances.

**Reply by jccalvojackson** [reply]
thank you!
